### PR TITLE
S3 cse encryption and compression

### PR DESCRIPTION
**Pull request by natesammons-nasdaq**
This change adds support for gzip compression on data stored in S3 to reduce storage size.

This change also adds support for S3 encryption materials provider implementations when dealing with data in S3.  This allows for client-managed encryption keys which are still transparently used when viewing/downloading output files stored in S3.

### CODE REVIEW

File: src/main/java/com/airbnb/airpal/modules/AirpalModule.java [hunk 1]
```diff
@@ -270,12 +272,40 @@ public AWSCredentials provideAWSCredentials()
...
+                    emp = (EncryptionMaterialsProvider)instance;
+                }
+                else {
+                    throw new IllegalArgumentException("Class " + empClassName + " must implement EncryptionMaterialsProvider");
+                }
+            }
+            catch (Exception x) {
+                throw new RuntimeException("Unable to initialize EncryptionMaterialsProvider class " + empClassName + ": " + x, x);
+            }
+        }
```
Comment (andykram) [inline src/main/java/com/airbnb/airpal/modules/AirpalModule.java]: I think we want to move the `EncryptionMaterialsProvider` into its own provider method, having this method depend on it.

File: src/main/java/com/airbnb/airpal/resources/ResultsPreviewResource.java [hunk 2]
```diff
@@ -82,7 +84,9 @@ private Response getS3Preview(URI fileURI, int numLines) {
                 outputKey
         ).withRange(0, 100 * 1024);
         val object = s3Client.getObject(request);
-        try (val s3Reader = new CSVReader(new BufferedReader(new InputStreamReader(object.getObjectContent())))) {
+        ObjectMetadata objectMetadata = object.getObjectMetadata();
+        boolean gzip = "gzip".equalsIgnoreCase(objectMetadata.getContentEncoding());
+        try (val s3Reader = new CSVReader(new BufferedReader(new InputStreamReader(gzip ? new GZIPInputStream(object.getObjectContent()) : object.getObjectContent())))) {
```
Comment (andykram) [inline src/main/java/com/airbnb/airpal/resources/ResultsPreviewResource.java]: Please break this line up a bit, it's a bit too complicated as it stands. Simply moving the result of the ternary to the result of a multiline `if..else` would be sufficient.

File: src/main/java/com/airbnb/airpal/api/output/builders/CsvOutputBuilder.java [hunk 3]
```diff
@@ -34,13 +35,13 @@
...
 
-    public CsvOutputBuilder(boolean includeHeader, UUID jobUUID, long maxFileSizeBytes) throws IOException {
+    public CsvOutputBuilder(boolean includeHeader, UUID jobUUID, long maxFileSizeBytes, boolean compressedOutput) throws IOException {
         this.includeHeader = includeHeader;
         this.jobUUID = jobUUID;
         this.outputFile = File.createTempFile(jobUUID.toString(), FILE_SUFFIX);
         this.maxFileSizeBytes = maxFileSizeBytes;
         this.countingOutputStream = new CountingOutputStream(new FileOutputStream(this.outputFile));
-        this.csvWriter = new CSVWriter(new OutputStreamWriter(this.countingOutputStream));
+        this.csvWriter = new CSVWriter(new OutputStreamWriter(compressedOutput ? new GZIPOutputStream(this.countingOutputStream) : this.countingOutputStream));
```
Comment (andykram) [inline src/main/java/com/airbnb/airpal/api/output/builders/CsvOutputBuilder.java]: Please break this line up like described in https://github.com/airbnb/airpal/pull/160/files#r51830841

**Comment by andykram** [comment]
Sorry for taking a while to get to this PR, but I left a few comments. If you can address these comments I'd be more than happy to merge your pr in

**Comment by natesammons-nasdaq** [comment]
Cool — thanks.  I'll try and get to this this week, but all the comments look good.

Thanks

-nate

Nate Sammons
Principal Architect
[cid:image002.png@01CFDD78.1CD32F50]

**Comment by natesammons-nasdaq** [comment]
OK, just pushed changes – decided this was more interesting than my conference call this morning :)

-nate

Nate Sammons
Principal Architect
[cid:image002.png@01CFDD78.1CD32F50]

**Comment by andykram** [comment]
Looks great, thanks for your contribution @natesammons-nasdaq

**Comment by natesammons-nasdaq** [comment]
Thanks!  This is great -- thanks for open sourcing this tool, it's extremely handy.

Nate

Sent from my iPhone

On Feb 4, 2016, at 12:29 PM, Andy Kramolisch <notifications@github.com<mailto:notifications@github.com>> wrote:


Looks great, thanks for your contribution @natesammons-nasdaq<https://github.com/natesammons-nasdaq>

—
Reply to this email directly or view it on GitHub<https://github.com/airbnb/airpal/pull/160#issuecomment-180014161>.
***********************************************************
CONFIDENTIALITY NOTICE: This e-mail and any attachments are for the exclusive and confidential use of the intended recipient and may constitute non-public information. If you received this e-mail in error, disclosing, copying, distributing or taking any action in reliance of this e-mail is strictly prohibited and may be unlawful. Instead, please notify us immediately by return e-mail and promptly delete this message and its attachments from your computer system. We do not waive any work product or other applicable legal privilege(s) by the transmission of this message.
***********************************************************

**Pull request merged**